File: frontend/src/types/message.ts
/**
 * Direct messaging types.
 */

export interface Message {
  id: string
  conversation_id: string
  sender_id: string
  content: string
  created_at: string
  read_at: string | null
}

export interface LastMessage {
  id: string
  content: string
  sender_id: string
  created_at: string
}

export interface Conversation {
  conversation_id: string
  friend_id: string
  friend_display_name: string | null
  friend_avatar_url: string | null
  is_online: boolean
  last_message: LastMessage | null
  unread_count: number
  updated_at: string
}

export interface ConversationListResponse {
  conversations: Conversation[]
  total_unread: number
}

export interface MessageHistoryResponse {
  messages: Message[]
  has_more: boolean
  oldest_id: string | null
}

// WebSocket event payloads
export interface DMMessagePayload {
  message_id: string
  conversation_id: string
  sender_id: string
  sender_display_name: string | null
  content: string
  created_at: string
}

export interface DMReadPayload {
  conversation_id: string
  reader_id: string
}

File: backend/app/schemas/message.py
"""
Message schemas for direct messaging.
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class SendMessageRequest(BaseModel):
    """Request to send a message."""
    content: str = Field(..., min_length=1, max_length=500)


class MessageResponse(BaseModel):
    """A single message."""
    id: str
    conversation_id: str
    sender_id: str
    content: str
    created_at: datetime
    read_at: Optional[datetime] = None


class LastMessage(BaseModel):
    """Last message preview for conversation list."""
    id: str
    content: str
    sender_id: str
    created_at: datetime


class ConversationResponse(BaseModel):
    """A conversation with metadata."""
    conversation_id: str
    friend_id: str
    friend_display_name: Optional[str] = None
    friend_avatar_url: Optional[str] = None
    is_online: bool = False
    last_message: Optional[LastMessage] = None
    unread_count: int = 0
    updated_at: datetime


class ConversationListResponse(BaseModel):
    """List of conversations."""
    conversations: list[ConversationResponse]
    total_unread: int = 0


class MessageHistoryResponse(BaseModel):
    """Paginated message history."""
    messages: list[MessageResponse]
    has_more: bool = False
    oldest_id: Optional[str] = None
